src/Dialect/ONNX/Transforms/Decompose.cpp

```diff
@@ -2826,59 +2828,465 @@ struct CustomOpFuseMatMulPattern : public OpRewritePattern<ONNXCustomOp> {
   }
 };
 
-namespace {
+static constexpr StringLiteral MicrosoftDomainName("com.microsoft");
+static constexpr StringLiteral DefaultONNXDomainName("");
 
-[[nodiscard]] bool isCustomMicrosoftOp(
-    ONNXCustomOp customOp, StringRef expectedName) {
+[[nodiscard]] bool isCustomOpWithNameAndDialect(
+    ONNXCustomOp customOp, StringRef expectedName, StringRef expectedDialect) {
   if (!customOp.getFunctionName().equals_insensitive(expectedName)) {
     return false;
   }
 
   const auto domAttr = customOp->getAttrOfType<StringAttr>("domain_name");
-  return domAttr && domAttr.getValue().equals_insensitive("com.microsoft");
+  return domAttr && domAttr.getValue().equals_insensitive(expectedDialect);
 }
 
-} // namespace
-
-template <typename OpToCreate>
-struct CustomOpMicrosoftToOnnxOp : public OpRewritePattern<ONNXCustomOp> {
-  CustomOpMicrosoftToOnnxOp(MLIRContext *context,
-      std::string operationNameToRewrite, PatternBenefit benefit = 1)
-      : OpRewritePattern<ONNXCustomOp>(context, benefit),
-        operationNameToRewrite(std::move(operationNameToRewrite)) {}
+struct CustomOpToOnnxOps : public OpRewritePattern<ONNXCustomOp> {
+  CustomOpToOnnxOps(MLIRContext *context, StringRef dialect,
+      StringRef operationNameToRewrite, PatternBenefit benefit = 1)
+      : OpRewritePattern<ONNXCustomOp>(context, benefit), dialect(dialect),
+        operationNameToRewrite(operationNameToRewrite) {}
 
   LogicalResult matchAndRewrite(
       ONNXCustomOp customOp, PatternRewriter &rewriter) const final {
-    if (!isCustomMicrosoftOp(customOp, operationNameToRewrite)) {
+    if (!isCustomOpWithNameAndDialect(
+            customOp, operationNameToRewrite, dialect)) {
+      return failure();
+    }
+
+    return matchAndRewriteImpl(customOp, rewriter);
+  }
+
+  virtual LogicalResult matchAndRewriteImpl(
+      ONNXCustomOp /*customOp*/, PatternRewriter & /*rewriter*/) const {
+    return failure();
+  }
+
+  static LogicalResult verifyOpValidity(Operation *op) {
+    assert(op);
+    onnx_mlir::IgnoreDiagnostic diag(op->getContext()->getDiagEngine());
+    if (auto info = op->getName().getRegisteredInfo()) {
+      return info->verifyInvariants(op);
+    }
+    return mlir::verify(op);
+  }
+
+  static LogicalResult verifyOpsErasingOnError(
+      ValueRange values, PatternRewriter &rewriter) {
+    if (llvm::any_of(values, [](Value value) {
+          return value && failed(verifyOpValidity(value.getDefiningOp()));
```

Comment:
isn't value always valid?
```suggestion
          return failed(verifyOpValidity(value.getDefiningOp()));
```

Reply:
No, it can be null. I use this to handle optional arguments/results